src/transformers/trainer.py

```diff
@@ -233,6 +242,83 @@ def __init__(
         if self.args.fp16 and _use_native_amp:
             self.scaler = torch.cuda.amp.GradScaler()
 
+    @classmethod
+    def from_nlp_dataset(
+        cls,
+        model: PreTrainedModel,
+        args: TrainingArguments,
+        dataset: "nlp.dataset_dict.DatasetDict",
```

Comment:
Think it would prefer to not have dataset as an input argument but `train_dataset` and `eval_dataset`. IMO, this has two advantages:
1) The user is not forced to use the whole dataset (e.g. if he wants to just evaluate it would be much less costly to just use the eval set for RAM) 
2) The class method signature stays the same then the init signature

also putting @lhoestq in cc here

Reply:
Alternative: we can support both by having `dataset_dict`, `train_dataset` and `eval_dataset` as arguments (with an assert to check that not both are provided at the same type).

Reply:
Is there a big advantage to having `dataset_dict` as an input? It saves a couple of lines of code, but I don't really see a big argument for it

Reply:
The most common use case of Trainer is train + eval, and the preprocessing is often the same for both training and evaluation, so IMO it's more useful to support `dataset_dict`.

If we go down the road of "it just saves a few lines of code", then we can just say there is no need for a class method and the user can use the regular init ;-)

Reply:
I think I would go with the more flexible `train_dataset` and `eval_dataset`, in particular since the internal names for these in a dict might be different (in particular `test` can sometimes be `validation`).

It doesn't even save you a line of code in most cases (you just split the dict in the call) so better to have it explicit imo.

Reply:
I agree that from the user's perspective it can be confusing to see that the input expects a dataset dict. Asking explicitly for the train/val/test datasets splits is clearer imo